alacritty/src/renderer/builtin_font.rs

```diff
@@ -0,0 +1,751 @@
+        let steep = (to_y - from_y).abs() > (to_x - from_x).abs();
+        if steep {
+            mem::swap(&mut from_x, &mut from_y);
+            mem::swap(&mut to_x, &mut to_y);
+        }
+        if from_x > to_x {
+            mem::swap(&mut from_x, &mut to_x);
+            mem::swap(&mut from_y, &mut to_y);
+        }
+
+        let delta_x = to_x - from_x;
+        let delta_y = to_y - from_y;
+        let gradient = if delta_x.abs() <= f32::EPSILON { 1. } else { delta_y / delta_x };
+
+        let x_end = f32::round(from_x);
+        let y_end = from_y + gradient * (x_end - from_x);
+        let x_gap = 1. - (from_x + 0.5).fract();
+
+        let xpxl1 = x_end;
+        let ypxl1 = y_end.trunc();
+
+        let color_1 = Pixel::gray(((1. - y_end.fract()) * x_gap * COLOR_FILL._r as f32) as u8);
+        let color_2 = Pixel::gray((y_end.fract() * x_gap * COLOR_FILL._r as f32) as u8);
+        if steep {
+            self.put_pixel(ypxl1, xpxl1, color_1);
+            self.put_pixel(ypxl1 + 1., xpxl1, color_2);
+        } else {
+            self.put_pixel(xpxl1, ypxl1, color_1);
+            self.put_pixel(xpxl1 + 1., ypxl1, color_2);
+        }
+
+        let mut intery = y_end + gradient;
+
+        let x_end = f32::round(to_x);
+        let y_end = to_y + gradient * (x_end - to_x);
+        let x_gap = (to_x + 0.5).fract();
+        let xpxl2 = x_end;
+        let ypxl2 = y_end.trunc();
+
+        let color_1 = Pixel::gray(((1. - y_end.fract()) * x_gap * COLOR_FILL._r as f32) as u8);
+        let color_2 = Pixel::gray((y_end.fract() * x_gap * COLOR_FILL._r as f32) as u8);
+        if steep {
+            self.put_pixel(ypxl2, xpxl2, color_1);
+            self.put_pixel(ypxl2 + 1., xpxl2, color_2);
+        } else {
+            self.put_pixel(xpxl2, ypxl2, color_1);
+            self.put_pixel(xpxl2, ypxl2 + 1., color_2);
+        }
+
+        if steep {
+            for x in xpxl1 as i32 + 1..xpxl2 as i32 {
+                let color_1 = Pixel::gray(((1. - intery.fract()) * COLOR_FILL._r as f32) as u8);
+                let color_2 = Pixel::gray((intery.fract() * COLOR_FILL._r as f32) as u8);
+                self.put_pixel(intery.trunc(), x as f32, color_1);
+                self.put_pixel(intery.trunc() + 1., x as f32, color_2);
+                intery += gradient;
+            }
+        } else {
+            for x in xpxl1 as i32 + 1..xpxl2 as i32 {
+                let color_1 = Pixel::gray(((1. - intery.fract()) * COLOR_FILL._r as f32) as u8);
+                let color_2 = Pixel::gray((intery.fract() * COLOR_FILL._r as f32) as u8);
+                self.put_pixel(x as f32, intery.trunc(), color_1);
+                self.put_pixel(x as f32, intery.trunc() + 1., color_2);
+                intery += gradient;
+            }
+        }
+    }
+
+    /// Draws a part of ellipse centered in `(0., 0.)` with `self.x_center()` and `self.y_center`
+    /// radius with a given `stroke` in second quadrant of the coordinate system.
```

Comment:
```suggestion
    /// Draws a part of an ellipse centered in `(0., 0.)` with `self.x_center()` and `self.y_center()`
    /// radius with a given `stroke` in top-right quadrant of the coordinate system.
```

Second is a poor way to describe a quadrant because the numbering is somewhat subjective.

Also naming two values for a radius is kinda confusing and afaik not really the proper terminology for an ellipse. One could use the "proper" terms which are vertex/co-vertex according to Wikipedia but I don't think that's particularly helpful. Just saying width/height might be the easiest to understand?

Reply:
The thing is that it's not a **top-right** in our case, since our y is towards bottom. Actually quadrants are defined and the first one is where both signs are `+`, and using `vertex` and `co-vertex`, so using proper terminology should be fine.

Reply:
> so using proper terminology should be fine.

Strongly disagree. Having some "proper" terminology only makes things easier to look up, it doesn't make them intuitive. Using the direction immediately makes things intuitive, there's no need to look anything up when it's already obvious due to its simple phrasing.

Yes quadrants might be well-defined, but that's already making so many assumptions about the space we're working in (mathematics), the education of the reader (has it been taught to you properly?), the accuracy of the author (did you actually remember it correctly?). There's just so many unnecessary points of conflict.